octodns/provider/mythicbeasts.py

```diff
@@ -0,0 +1,420 @@
+    assert value[-1] == '.', 'Value already missing trailing dot'
+    return value[:-1]
+
+
+class MythicBeastsProvider(BaseProvider):
+    '''
+    Mythic Beasts DNS API Provider
+
+    mythicbeasts:
+      class: octodns.provider.mythicbeasts.MythicBeastsProvider
+        zones:
+          my-zone: 'password'
+    '''
+
+    SUPPORTS_GEO = False
+    SUPPORTS_DYNAMIC = False
+    SUPPORTS = set(('A', 'AAAA', 'ALIAS', 'CNAME', 'MX', 'NS',
+                    'SRV', 'SSHFP', 'CAA', 'TXT'))
+    BASE = 'https://dnsapi.mythic-beasts.com/'
+
+    def __init__(self, identifier, passwords, *args, **kwargs):
+        self.log = getLogger('MythicBeastsProvider[{}]'.format(identifier))
+
+        assert isinstance(passwords, dict), 'Missing passwords'
+
+        self.log.debug(
+            '__init__: id=%s, registered zones; %s',
+            identifier,
+            passwords.keys())
+        super(MythicBeastsProvider, self).__init__(identifier, *args, **kwargs)
+
+        self._passwords = passwords
+        sess = Session()
+        self._sess = sess
+
+    def _request(self, method, path, data=None):
+        self.log.debug('_request: method=%s, path=%s data=%s',
+                       method, path, data)
+
+        resp = self._sess.request(method, path, data=data)
+        self.log.debug(
+            '_request:   status=%d data=%s',
+            resp.status_code,
+            resp.text[:20])
+
+        if resp.status_code == 401:
+            raise Exception('Mythic Beasts unauthorized for domain: {}'
+                            .format(data['domain']))
+        resp.raise_for_status()
+        return resp
+
+    def _post(self, data=None):
+        return self._request('POST', self.BASE, data=data)
+
+    def records(self, zone):
+        assert zone in self._passwords, 'Missing password for domain: {}' \
+            .format(remove_trailing_dot(zone))
+
+        return self._post({
+            'domain': remove_trailing_dot(zone),
+            'password': self._passwords[zone],
+            'showall': 0,
+            'command': 'LIST',
+        })
+
+    @staticmethod
+    def _data_for_single(_type, data):
+        return {
+            'type': _type,
+            'value': data['raw_values'][0]['value'],
+            'ttl': data['raw_values'][0]['ttl']
+        }
+
+    @staticmethod
+    def _data_for_multiple(_type, data):
+        return {
+            'type': _type,
+            'values':
+                [raw_values['value'] for raw_values in data['raw_values']],
+            'ttl':
+                max([raw_values['ttl'] for raw_values in data['raw_values']]),
+        }
+
+    @staticmethod
+    def _data_for_MX(_type, data):
+        ttl = max([raw_values['ttl'] for raw_values in data['raw_values']])
+        values = []
+
+        for raw_value in \
+                [raw_values['value'] for raw_values in data['raw_values']]:
+            match = re.match('^([0-9]+)\\s+(\\S+)$', raw_value, re.IGNORECASE)
```

Comment:
Would prefer that this use a pre-compiled re rather than match which builds one each route. Pedantic and this isn't going to be a hot path or anything, but making it a class-static property, `MX_RE` etc., would compile it once and be more in line with other providers.

It's also best to use `r'...'` for regexes.

Reply:
Okay, I've changed that although I'm pretty sure python compiles and caches re.match etc even if you use it multiple times. I recall doing a test in 2.7 many months ago wondering if there was a difference but couldn't see any between re.compile and loop, vs, loop with re.match in the middle